### PR TITLE
Prevent sticking of filament during unload process

### PR DESCRIPTION
**Pull request by xKoldx**
Extrudes 3mm of a filament to "unstick" filament from hotend, then continues with unloading process. 

I've been having issues during filament change where the filament will get stuck and not back out unless I first "load" for a short period of time, then "unload" right after.

Please note the comment of 5mm is incorrect, forgot to change it back to 3mm.
5mm was too much.

### CODE REVIEW

**Comment by andrivet** [comment]
Never had this kind of problem. This is seems very specific to your hardware or the way you are working. The idea to extrude filament during unloading (i.e. un-extruding) is strange. I am sure if I implement your idea that several people will report a bug.

**Comment by xKoldx** [comment]
Good point. I'll have to look into it as I've upgraded to the microswiss. If I remember correctly I noticed that as a feature of Silverquarks I3 plus plus build. That's where I got the idea. But your right, at first I thought it was a bug but just left it alone. Switching to your build it no longer did it.

**Comment by chschmit** [comment]
Interesting. On mine the filament used to retract perfectly until I switched to the Microswiss. Now all firmwares have been problematic. So if I initiate the remove filament sequence I wait until it hits the temperature and then hold done the extruder lever so can manually push the filament into the hotend a few millimeters and let go of the lever. It will then back out the filament. Otherwise the extruder gear will slip.

On June 16, 2018 12:38:05 PM PDT, Dylan <notifications@github.com> wrote:
>Good point. I'll have to look into it as I've upgraded to the
>microswiss. If I remember correctly I noticed that as a feature of
>Silverquarks I3 plus plus build. That's where I got the idea. But your
>right, at first I thought it was a bug but just left it alone.
>Switching to your build it no longer did it. 
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/andrivet/ADVi3pp-Marlin/pull/146#issuecomment-397834716

-- 
Sent from my Android Google Pixel.

**Comment by xKoldx** [comment]
I know for a fact I followed their directions perfectly with the position of the heat break, And that the spacing was exact. I don't really know why it wouldn't heat high enough up to melt the filament that "expanded" near the top.

**Comment by SureshotM6** [comment]
Even with the stock extruder, I've had some poor unloads using the current implementation of the LCD unload functionality, especially with PETG.  Generally, the extruder drive gear usually grinds the filament for little bit at first (including on load), and then leaves more filament than desired behind in the nozzle when it finally does unload.  I think part of it is due to the unload starting when the thermistor gets within 10 degrees of the target, which may not be enough to melt the filament.  Additionally, I think the thermistor may register a higher temperature than the nozzle while heating, and some lag time may be required for the nozzle/filament to reach the same temperature, which I believe is part of the reason why Marlin has its TEMP_RESIDENCY_TIME setting.

I've been playing around with some modifications to the unload code myself for the past couple of months.  I've personally had better luck (no grinding and more extraction) with preheating all the way to the target temperature (not to target -10 degrees C), extruding, and then retracting, but I don't switch filaments very often so it hasn't had extensive testing.  I've noticed some other firmwares extract some filament first during unload, like this commit, to "pressurize" the nozzle. which seems to make sense.

So basically, I think this commit would be an improvement, but I think other changes would be good as well.  Once I'm happy with the changes I've been making, I can submit a pull request or feature request with info as well.  I could be wrong though -- the amount of filament left behind may be comparable.

**Comment by xKoldx** [comment]
Well its good to know that im not the only one, and im not crazy! lol. 

Thanks for both of your guys' input.

**Comment by LordJakson** [comment]
It happens to me a lot of times with the stock extruder and PLA.
I usually press "Unload" and wait until the filament comes out or the extruder drive gear has grinded the filament.
In this case I hold down the lever and feed some filament manualy until the drive gear has new grip and the filament comes out.
